### PR TITLE
Support linking of tasks together

### PR DESCRIPTION
**Pull request by oliviermaridat**
As described in  #92, this pull request adds the ability to link too tasks together in order to be able to explain the relation between too tasks and therefore organize better the work. This also simplify the browsing between tasks.

* A list of link labels per project can be managed through the Project view
* A list of default link labels for new project can be managed through the Config view (Board settings)
* A link to an other task can be added in the Task show view
 * Linking two tasks of the same project is the common behaviour
 * Linking two tasks of different project is working, but only the link labels of the current project are available during the link creation
* The number of links to other tasks is displayed at the bottom of each task in the Board view
* Project links are duplicated when a project is duplicated
* Task links are duplicated when a task is duplicated
* This PR has been tested for SQLite and MySQL but not for Postgres
* This PR includes English and French translations only

### CODE REVIEW

**Comment by oliviermaridat** [comment]
Sorry for the high number of commits, I don't know (yet) how to handle that more properly.

**Comment by fguillot** [comment]
Awesome, give me the week to test and review that.

File: index.php [hunk 1]
```diff
@@ -1,5 +1,7 @@
 <?php
-
+ini_set('display_errors',1);
+ini_set('display_startup_errors',1);
+error_reporting(-1);
```
Comment (fguillot) [inline index.php]: error_reporting should not be hardcoded

File: app/Template/tasklink/edit.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+    <div class="form-actions">
+        <input type="submit" value="<?= t('Save') ?>" class="btn btn-blue"/>
+        <?= t('or') ?>
+        <?= $this->a(t('cancel'), 'task', 'show', array('task_id' => $task['id'], 'project_id' => $task['project_id'])) ?>
+    </div>
+</form>
+<?php else: ?>
+<div class="alert alert-info">
+    You need to <?= $this->a('add link labels', 'link', 'index', array('project_id' => $task['project_id'])) ?> to this project before to link this task to another one.
```
Comment (fguillot) [inline app/Template/tasklink/edit.php]: This text cannot be translated.

File: app/Template/link/index.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+		<i><?= t('#10 Follows #9') ?></i>
+	</div>
+    <?= $this->formCsrf() ?>
+    <?= $this->formHidden('project_id', $values) ?>
+
+    <?= $this->formLabel(t('Link Label'), 'name') ?>
+    <?= $this->formText('name', $values, $errors, array('required', 'autofocus', 'placeholder="Precedes"')) ?>
+
+    <?= $this->formLabel(t('Link Inverse Label'), 'name_inverse') ?>
+    <?= $this->formText('name_inverse', $values, $errors, array('required', 'placeholder="Follows"')) ?>
```
Comment (fguillot) [inline app/Template/link/index.php]: placeholders are not translated

File: app/Model/TaskLink.php [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+                    return false;
+                }
+            }
+        });
+    }
+
+    public function isUniqueConstraintFailed()
+    {
+        return $this->search($this->db->getLogMessages(), 'UNIQUE') || $this->search($this->db->getLogMessages(), 'Duplicata');
+    }
```
Comment (fguillot) [inline app/Model/TaskLink.php]: You should not do that. That will work only in debug mode. Check if the link exists at the validation.

File: app/Controller/Tasklink.php [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+                'task_id' => $task['id'],
+                'another_link' => $this->request->getIntegerParam('another_link', 0)
+            );
+        }
+
+        $this->response->html($this->taskLayout('tasklink/create', array(
+            'values' => $values,
+            'errors' => $errors,
+            'link_list' => $this->link->getList($task['project_id'], false),
+            'task_list' => $this->taskFinder->getList($task['project_id'], Task::STATUS_OPEN, $task['id']),
```
Comment (fguillot) [inline app/Controller/Tasklink.php]: I got a fatal error here, should be "TaskModel::STATUS_OPEN"

**Comment by fguillot** [comment]
Why the links are specific per project, why not global to the application? That simplify a lot of stuff. 

Categories and swimlanes are specific per project because they can be different but you will always define the same link types "blocked by/blocks this task...", it's like task colors for me.

**Comment by oliviermaridat** [comment]
Thank you very much for all your comments. I checked and checked my code, but apparently not enough. I will see that asap.

> Why the links are specific per project, why not global to the application?

I made them specific per project because I thought they may be different depending of business activities (online shops project will have different links than development project) like categories. So, not exactly like task colors in my opinion. 
But for sure this add more complexity. This is your call, what do you prefer?

**Comment by fguillot** [comment]
Jira have those links:

link_jira

I think those "relations" should be there by default, that cover most needs.

I would prefer to have links global to the application, by example available from a new section in the settings page.

My concern is you have done a lot of works to make your pull-request, are you ok to change that? Maybe someone else have a different opinion? after all it's an open source software.

**Comment by oliviermaridat** [comment]
Normally I have fixed all the issues you spotted. But the PR will fail, because I am using my proposed PR for [simpleValidator](https://github.com/fguillot/simpleValidator) (not to force your hand of course, task link validation can be done differently in kanboard if you prefer). 

This will give me the time to think about "global links" ;-) I am very open and I am ok to change whatever you want. I just want to do it the best possible way.

**Comment by oliviermaridat** [comment]
I am working in a different branch ([oliviermaridat/kanboard/link](https://github.com/oliviermaridat/kanboard/tree/link)) to make link labels global to all project. Actually, this doesn't change many things in the code, but it will clearly simplify the management for the end-user. Anyway, it will still be possible to add a "link label" for a specific project. This offers simplicity and specificity :-) But if you don't like "specificity", feel free to remove it.
My only regrets is that I can't add that properly to the Controller/Config (too much pages to create), so I keep it in Controller/Link even it is accessible through the application Settings page.

I am also working on clarifying the "link" and "link label" data models. I was never completely satisfied with it, but I didn't find how to do that properly. I have a new idea, so I will try it.

I will send a new PR when this will be done.

**Comment by BlueTeck** [comment]
I like it :+1: 

Note: The default links should be translated.

**Comment by fguillot** [comment]
I close this pull-request, just open a new one when you will be ready.

Thanks.

**Pull request closed without merging**